src/libmongoc/src/mongoc/mongoc-cyrus.c

```diff
@@ -126,6 +126,49 @@ _mongoc_cyrus_get_user (mongoc_cyrus_t *sasl, int param_id, const char **result,
    return (sasl->credentials.user != NULL) ? SASL_OK : SASL_FAIL;
 }
 
+static const char *
+sasl_verify_type_to_str (sasl_verify_type_t type)
+{
+   switch (type) {
+   case SASL_VRFY_PLUGIN:
+      return "SASL_VRFY_PLUGIN";
+   case SASL_VRFY_CONF:
+      return "SASL_VRFY_CONF";
+   case SASL_VRFY_PASSWD:
+      return "SASL_VRFY_PASSWD";
+   case SASL_VRFY_OTHER:
+      return "SASL_VRFY_OTHER";
+   default:
+      return "Unknown";
+   }
+}
+
+int
+_mongoc_cyrus_verifyfile_cb (void *context, const char *file, sasl_verify_type_t type)
+{
+   TRACE ("Attempting to load file: `%s`. Type is %s\n", file, sasl_verify_type_to_str (type));
+
+#ifdef _WIN32
+   // On Windows, Cyrus-SASL hard-codes the plugin path.
+   // Only permit loading plug-in from user configured path to prevent unintentional library loading.
+   if (type == SASL_VRFY_PLUGIN) {
+      const char *path_prefix = MONGOC_CYRUS_PLUGIN_PATH_PREFIX;
+      bool has_valid_prefix = (0 != strlen (path_prefix) && file == strstr (file, path_prefix));
+      // Check if `file` has necessary prefix.
+      if (has_valid_prefix) {
+         return SASL_OK;
+      }
+      MONGOC_WARNING ("Not loading Cyrus-SASL plugin: %s. If plugin is needed, set CMake option "
+                      "`CYRUS_PLUGIN_PATH_PREFIX (currently '%s')` to a non-empty string contain the path prefix.",
```

Comment:
```suggestion
      MONGOC_WARNING ("Refusing to load Cyrus-SASL plugin at: '%s'. If needed, set CYRUS_PLUGIN_PATH_PREFIX (currently "
                      "'%s') to the absolute path prefix during build configuration.",
```
Message phrasing suggestion.